** Standard scaler**

In [1]:
import numpy as np
import pandas as pd

C:\Users\Admin\AppData\Local\Temp\ipykernel_16532\1662815981.py:2: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [2]:
import warnings
warnings.filterwarnings('ignore')


In [3]:
Dataset1 = pd.read_csv(r"C:\Users\Admin\Desktop\Viska\Machine Learning\Day-11 standard scaler((recording in Power BI 11th May'24))/diabetes.csv")
Dataset1

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

# here we can the value of Glucose column is very high compare to other columns. 
# model give more importance to glucose column compare to ther column. but every column are equally important.
# so here model will became bias and didn't give accurate result.
# for that here we have to use standard scaler.
# if difference is less then biasness also less and model give accurate result.

In [4]:
Dataset1.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [5]:
#Splitting dataset into input and output
x= Dataset1.drop("Outcome",axis=1)
x

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  
0                       0.627   50  
1                       0.351   31  
2                       0.672   32  
3                       0.167   21  
4                       2.288   33  
..                        ...  ...  
763                     0.171   63  
764                     0.340   27  
765                     0.245   30  
766                     0.349   47  
767                     0.315   23  

[768 rows x 8 columns]

In [6]:
y=Dataset1.loc[:,['Outcome']]
y

Outcome
0          1
1          0
2          1
3          0
4          1
..       ...
763        0
764        0
765        0
766        1
767        0

[768 rows x 1 columns]

In [7]:
#Train_test_split
from sklearn.model_selection import train_test_split

In [8]:
x_train,x_test,y_train,y_test = train_test_split(x,y,train_size=0.75,random_state=42)

In [9]:
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(576, 8)
(192, 8)
(576, 1)
(192, 1)


In [10]:
#shape of my original dataset
print(Dataset1.shape)

(768, 9)


# Applying the standard scaler
# standard scaler only apply on input data (independent data)
# apply on x_train

In [11]:
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()

In [12]:
#Learning on x_train
sc.fit(x_train)

StandardScaler()

In [13]:
#Transform to SC -X_train,X_test
x_train_sc=sc.transform(x_train)
x_test_sc=sc.transform(x_test)


In [14]:
#Original x_train
x_train

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
357           13      129              0             30        0  39.9   
73             4      129             86             20      270  35.1   
352            3       61             82             28        0  34.4   
497            2       81             72             15       76  30.1   
145            0      102             75             23        0   0.0   
..           ...      ...            ...            ...      ...   ...   
71             5      139             64             35      140  28.6   
106            1       96            122              0        0  22.4   
270           10      101             86             37        0  45.6   
435            0      141              0              0        0  42.4   
102            0      125             96              0        0  22.5   

     DiabetesPedigreeFunction  Age  
357                     0.569   44  
73                      0.231   23  
352                     0.243   46  
497                     0.547   25  
145                     0.572   21  
..                        ...  ...  
71                      0.411   26  
106                     0.207   27  
270                     1.136   38  
435                     0.205   29  
102                     0.262   21  

[576 rows x 8 columns]

In [15]:
#Transform X_train_sc
pd.DataFrame(x_train_sc)

0         1         2         3         4         5         6  \
0    2.803468  0.259779 -3.780779  0.616770 -0.692052  1.039740  0.296085   
1    0.078327  0.259779  0.897245 -0.032106  1.633077  0.409454 -0.700876   
2   -0.224467 -1.858253  0.679662  0.486995 -0.692052  0.317537 -0.665480   
3   -0.527260 -1.235302  0.135706 -0.356544 -0.037571 -0.247095  0.231195   
4   -1.132847 -0.581204  0.298893  0.162557 -0.692052 -4.199517  0.304934   
..        ...       ...       ...       ...       ...       ...       ...   
571  0.381120  0.571254 -0.299459  0.941208  0.513571 -0.444059 -0.169950   
572 -0.830054 -0.768089  2.855487 -1.329858 -0.692052 -1.258179 -0.771666   
573  1.895088 -0.612352  0.897245  1.070984 -0.692052  1.788206  1.968502   
574 -1.132847  0.633549 -3.780779 -1.329858 -0.692052  1.368015 -0.777565   
575 -1.132847  0.135189  1.441201 -1.329858 -0.692052 -1.245048 -0.609438   

            7  
0    0.963521  
1   -0.862956  
2    1.137471  
3   -0.689006  
4   -1.036906  
..        ...  
571 -0.602031  
572 -0.515056  
573  0.441670  
574 -0.341105  
575 -1.036906  

[576 rows x 8 columns]

# here we can see the value are change

x_train_sc (new value) how it comes?
take first line of x_train

new value = original value - mean / standard deviation 
          = 13 - 3.741 / 3.305
          = 2.803

In [16]:
x_train.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   576.000000  576.000000     576.000000     576.000000  576.000000   
mean      3.741319  120.659722      69.505208      20.494792   80.362847   
std       3.305452   32.133182      18.399810      15.424655  116.223541   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      64.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   37.000000   
75%       6.000000  138.250000      80.000000      32.000000  127.500000   
max      17.000000  199.000000     122.000000      63.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age  
count  576.000000                576.000000  576.000000  
mean    31.981771                  0.468618   32.921875  
std      7.622203                  0.339325   11.507539  
min      0.000000                  0.078000   21.000000  
25%     27.000000                  0.240000   24.000000  
50%     32.000000                  0.372500   29.000000  
75%     36.325000                  0.612250   40.000000  
max     67.100000                  2.420000   81.000000

In [17]:
pd.DataFrame(x_train_sc)

0         1         2         3         4         5         6  \
0    2.803468  0.259779 -3.780779  0.616770 -0.692052  1.039740  0.296085   
1    0.078327  0.259779  0.897245 -0.032106  1.633077  0.409454 -0.700876   
2   -0.224467 -1.858253  0.679662  0.486995 -0.692052  0.317537 -0.665480   
3   -0.527260 -1.235302  0.135706 -0.356544 -0.037571 -0.247095  0.231195   
4   -1.132847 -0.581204  0.298893  0.162557 -0.692052 -4.199517  0.304934   
..        ...       ...       ...       ...       ...       ...       ...   
571  0.381120  0.571254 -0.299459  0.941208  0.513571 -0.444059 -0.169950   
572 -0.830054 -0.768089  2.855487 -1.329858 -0.692052 -1.258179 -0.771666   
573  1.895088 -0.612352  0.897245  1.070984 -0.692052  1.788206  1.968502   
574 -1.132847  0.633549 -3.780779 -1.329858 -0.692052  1.368015 -0.777565   
575 -1.132847  0.135189  1.441201 -1.329858 -0.692052 -1.245048 -0.609438   

            7  
0    0.963521  
1   -0.862956  
2    1.137471  
3   -0.689006  
4   -1.036906  
..        ...  
571 -0.602031  
572 -0.515056  
573  0.441670  
574 -0.341105  
575 -1.036906  

[576 rows x 8 columns]

In [18]:
 #Original x_test
x_test

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
668            6       98             58             33      190  34.0   
324            2      112             75             32        0  35.7   
624            2      108             64              0        0  30.8   
690            8      107             80              0        0  24.6   
473            7      136             90              0        0  29.9   
..           ...      ...            ...            ...      ...   ...   
554            1       84             64             23      115  36.9   
319            6      194             78              0        0  23.5   
594            6      123             72             45      230  33.6   
6              3       78             50             32       88  31.0   
615            3      106             72              0        0  25.8   

     DiabetesPedigreeFunction  Age  
668                     0.430   43  
324                     0.148   21  
624                     0.158   21  
690                     0.856   34  
473                     0.210   50  
..                        ...  ...  
554                     0.471   28  
319                     0.129   59  
594                     0.733   34  
6                       0.248   26  
615                     0.207   27  

[192 rows x 8 columns]

In [19]:
pd.DataFrame(x_test_sc)

0         1         2         3         4         5         6  \
0    0.683914 -0.705794 -0.625833  0.811433  0.944150  0.265013 -0.113907   
1   -0.527260 -0.269729  0.298893  0.746546 -0.692052  0.488240 -0.945691   
2   -0.527260 -0.394319 -0.299459 -1.329858 -0.692052 -0.155178 -0.916196   
3    1.289501 -0.425467  0.570871 -1.329858 -0.692052 -0.969298  1.142617   
4    0.986707  0.477812  1.114827 -1.329858 -0.692052 -0.273357 -0.762817   
..        ...       ...       ...       ...       ...       ...       ...   
187 -0.830054 -1.141860 -0.299459  0.162557  0.298281  0.645811  0.007026   
188  0.683914  2.284368  0.462080 -1.329858 -0.692052 -1.113739 -1.001734   
189  0.683914  0.072894  0.135706  1.590085  1.288613  0.212489  0.779818   
190 -0.224467 -1.328745 -1.060998  0.746546  0.065768 -0.128916 -0.650733   
191 -0.224467 -0.456614  0.135706 -1.329858 -0.692052 -0.811726 -0.771666   

            7  
0    0.876546  
1   -1.036906  
2   -1.036906  
3    0.093770  
4    1.485371  
..        ...  
187 -0.428081  
188  2.268147  
189  0.093770  
190 -0.602031  
191 -0.515056  

[192 rows x 8 columns]

x_test_sc (new value) how it comes?
take first line of x_test

new value = original value - mean / standard deviation 
          = 6 - 3.741 / 3.305
          = 0.683